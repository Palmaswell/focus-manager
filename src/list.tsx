import * as React from 'react';

export interface ListProps {
  tabIndex: number;
  onKeyPress?: () => void;
}

export class List extends React.Component<ListProps> {
  private readonly refList: React.RefObject<
    HTMLUListElement
  > = React.createRef();
  // private attachKeyPress = (c: HTMLUListElement) => {
  //   console.log('attach')
  //   c.addEventListener('keypress', () => {
  //     console.log('hi^^^^^')
  //     return this.props.onKeyPress || (() => {})
  //   })
  // }

  public render(): JSX.Element {
    const { tabIndex } = this.props;
    return (
      <ul ref={this.refList} role="listbox" tabIndex={tabIndex}>
        {this.props.children}
      </ul>
    );
  }
}

export interface ListItemProps {
  selected: boolean;
  children: any;
}

export function ListItem(props: ListItemProps): JSX.Element {
  return (
    <li aria-selected={props.selected} role="option" tabIndex={0}>
      {props.children}
    </li>
  );
}
