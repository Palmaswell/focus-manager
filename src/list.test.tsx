import * as React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';

import { List, ListItem } from './list';

afterEach(cleanup);

describe('Keyboard Manager', () => {
  test('should render without crashing', () => {
    const { getByRole } = render(
      <List tabIndex={0}>
        <ListItem selected={false}>Item One</ListItem>
        <ListItem selected={false}>Item Two</ListItem>
      </List>
    );
    expect(getByRole('listbox')).toMatchSnapshot();
    
  });
  test('First list element gets the focus', () => {
    const { getByText } = render(
      <List tabIndex={0}>
        <ListItem selected={false}>Item One</ListItem>
        <ListItem selected={false}>Item Two</ListItem>
      </List>
    );
    const firstLi = getByText('Item One');
    firstLi.focus();
    expect(document.activeElement).toBe(firstLi);
    expect(firstLi.getAttribute('aria-selected')).toBe(true);
  });
  test('OnKeydown the next element on the bottom gets the focus', () => {
    const { getByRole, getByText } = render(
      <List tabIndex={0}>
        <ListItem selected={false}>Item One</ListItem>
        <ListItem selected={false}>Item Two</ListItem>
      </List>
    );
    const firstLi = getByText('Item One');
    firstLi.focus();
    fireEvent.keyDown(getByRole('listbox'), { key: 'ArrowDown', code: 40 });

  });
});
