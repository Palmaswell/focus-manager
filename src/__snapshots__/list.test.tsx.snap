// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard Manager should render without crashing 1`] = `
<ul
  role="listbox"
  tabindex="0"
>
  <li
    aria-selected="false"
    role="option"
    tabindex="0"
  >
    Item One
  </li>
  <li
    aria-selected="false"
    role="option"
    tabindex="0"
  >
    Item Two
  </li>
</ul>
`;
